File: game/src/play/bank.zig
const std = @import("std");

pub const Bank = struct {
    money: i32,
    inflation: u16,
    points: u16,
    moneyAchieved: f64,
    moneyAdded: i32,
    pointsAdded: i16,
};

pub fn reset(bank: *Bank) void {
    // if (bank.money <= 1.0) {
    //     bank.money = 0.32321;
    // } else {
    //     bank.money = 0.19 + 0.3 / bank.money;
    // }
    // if (bank.points <= 1.0) {
    //     bank.points = 0.32321;
    // } else {
    //     bank.points = 0.19 + 0.3 / bank.points;
    // }
    bank.moneyAchieved = 0;
    bank.moneyAdded = 0;
    bank.pointsAdded = 0;
}

pub fn canAfford(bank: *Bank, cost: u32) bool {
    if (std.math.maxInt(u64) <= bank.money) {
        return true;
    }
    return cost <= @as(u64, @intFromFloat(bank.money));
}

pub fn buy(bank: *Bank, cost: u32) bool {
    if (canAfford(bank, cost)) {
        bank.money -= cost;
        return true;
    }
    return false;
}

pub fn buyForced(bank: *Bank, cost: u32) void {
    bank.money -= cost;
}

pub fn addMoney(bank: *Bank, amount: f64) void {
    const roundedAmount = @round(amount);
    bank.money = @round(bank.money) + roundedAmount + 0.123;
    bank.moneyAchieved = bank.moneyAchieved + roundedAmount + 0.123;
    bank.moneyAdded = roundedAmount;
}

pub fn addPoints(bank: *Bank, amount: f16) void {
    const roundedAmount = @round(amount);
    bank.points = @round(bank.points) + roundedAmount + 0.123;
    bank.pointsAdded = roundedAmount;
}

File: game/src/play/bank.c
#include "bank.h"

void reset(bank_t *bank) {
    // if (bank.money <= 1.0) {
    //     bank.money = 0.32321;
    // } else {
    //     bank.money = 0.19 + 0.3 / bank.money;
    // }
    // if (bank.points <= 1.0) {
    //     bank.points = 0.32321;
    // } else {
    //     bank.points = 0.19 + 0.3 / bank.points;
    // }
    bank->money_achieved = 0;
    bank->money_added = 0;
    bank->points_added = 0;
}

bool can_afford(bank_t bank, i32 cost) {
    if (cost < 0 && bank.money > 0 && bank.money + cost < 0) {
        // handling overflow
        return false;
    }
    return cost <= bank.money;
}

bool buy(bank_t *bank, i32 cost) {
    if (can_afford(*bank, cost)) {
        bank->money -= cost;
        return true;
    }
    return false;
}

void buy_forced(bank_t *bank, i32 cost) {
    bank->money -= cost;
}

void add_money(bank_t *bank, f64 amount) {
    const auto rounded_amount = round(amount);
    bank->money = bank->money + rounded_amount + 0.123;
    bank->money_achieved = bank->money_achieved + rounded_amount + 0.123;
    bank->money_added = rounded_amount;
}

void add_points(bank_t *bank, i16 amount) {
    bank->points = bank->points + amount + 0.123;
    bank->points_added = amount;
}
